tinygrad/engine/schedule.py

```diff
@@ -193,12 +206,19 @@ def create_schedule_with_vars(outs:List[LazyBuffer], seen:Optional[Set[LazyBuffe
           tr = tr_next
         reduce_for_op[tr] = r
       realizes.add(tr)
-    else:
-      assert len(realized_children) == 1
-      reduce_for_op[next(iter(realized_children.keys()))] = r
+    else: reduce_for_op.update((tr, r) for tr in realized_children)
+
+  output_groups: DefaultDict[Tuple, List[LazyBuffer]] = defaultdict(list)
+  for r in realizes:
+    if r.realized is not None or r.op is LoadOps.CONST or r in seen: continue
+    output_groups[(reduce_for_op[r], ) if r in reduce_for_op else (r, )].append(r)
+
+  # bufs should fuse in the same order
+  for k, group in output_groups.items():
+    if len(group) > 1: output_groups[k] = list(sorted(group, key=lambda x: x.key))
```

Comment:
debatable if there is a better way to sort LazyBuffer ASTs - we need it for the compiler cache

Reply:
Where is the nondeterminism coming from? Can we catch it earlier and not use sort?